File: bindings/nostr-sdk-ffi/bindings-python/examples/tor.py
import asyncio
from nostr_sdk import NostrSigner, Keys, Client, Options, EventBuilder, Connection, ConnectionTarget, init_logger, LogLevel
from datetime import timedelta


async def main():
    init_logger(LogLevel.INFO)

    keys = Keys.generate()
    print(keys.public_key().to_bech32())

    signer = NostrSigner.keys(keys)

    # Configure client to use embedded tor for `.onion` relays
    connection = Connection().embedded_tor().target(ConnectionTarget.ONION)
    opts = Options().connection(connection).connection_timeout(timedelta(seconds=60))
    client = Client.with_opts(signer, opts)

    await client.add_relays([
        "wss://relay.damus.io",
        "ws://oxtrdevav64z64yb7x6rjg4ntzqjhedm5b5zjqulugknhzr46ny2qbad.onion",
        "ws://2jsnlhfnelig5acq6iacydmzdbdmg7xwunm4xl6qwbvzacw4lwrjmlyd.onion",
    ])
    await client.connect()

    event = EventBuilder.text_note("Hello from rust-nostr Python bindings!", [])
    res = await client.send_event_builder(event)
    print("Event sent:")
    print(f" hex:    {res.id.to_hex()}")
    print(f" bech32: {res.id.to_bech32()}")
    print(f" Successfully sent to:    {res.output.success}")
    print(f" Failed to send to: {res.output.failed}")


if __name__ == '__main__':
    asyncio.run(main())

File: crates/nostr-sdk/examples/tor.rs
// Copyright (c) 2022-2023 Yuki Kishimoto
// Copyright (c) 2023-2024 Rust Nostr Developers
// Distributed under the MIT software license

use nostr_sdk::prelude::*;

const BECH32_SK: &str = "nsec1ufnus6pju578ste3v90xd5m2decpuzpql2295m3sknqcjzyys9ls0qlc85";

#[tokio::main]
async fn main() -> Result<()> {
    tracing_subscriber::fmt::init();

    // Parse keys
    let my_keys = Keys::parse(BECH32_SK)?;

    // Configure client to use embedded tor for `.onion` relays
    let connection: Connection = Connection::new()
        .embedded_tor()
        .target(ConnectionTarget::Onion);
    let opts = Options::new().connection(connection);
    let client = Client::with_opts(&my_keys, opts);

    // Add relays
    client
        .add_relays([
            "wss://relay.damus.io",
            "ws://oxtrdevav64z64yb7x6rjg4ntzqjhedm5b5zjqulugknhzr46ny2qbad.onion",
            "ws://2jsnlhfnelig5acq6iacydmzdbdmg7xwunm4xl6qwbvzacw4lwrjmlyd.onion",
        ])
        .await?;

    client.connect().await;

    let filter: Filter = Filter::new().pubkey(my_keys.public_key()).limit(0);
    client.subscribe(vec![filter], None).await?;

    // Handle subscription notifications with `handle_notifications` method
    client
        .handle_notifications(|notification| async {
            if let RelayPoolNotification::Event { event, .. } = notification {
                if event.kind == Kind::GiftWrap {
                    let UnwrappedGift { rumor, .. } = client.unwrap_gift_wrap(&event).await?;
                    println!("Rumor: {}", rumor.as_json());
                } else {
                    println!("{:?}", event);
                }
            }
            Ok(false) // Set to true to exit from the loop
        })
        .await?;

    Ok(())
}
